Features/108 matmul

coquelin77 opened the pull request:
implementation of matmul (and @ operator). If any changes are suggested/needed please let me know

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -690,6 +690,74 @@ def __eq__(self, other):
+            ----------
+            a : ht.DNDarray
+                2 dimensional: L x P
+
```
Extra line

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -690,6 +690,74 @@ def __eq__(self, other):
+
+            b : ht.DNDarray
+                2 dimensional: P x Q
+
```
Extra line

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -690,6 +690,74 @@ def __eq__(self, other):
+            out : ht.tensor
+                Optional
+                output tensor
+
```
Extra line

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -690,6 +690,74 @@ def __eq__(self, other):
+
+            References
+            ----------
+            [1] R. Gu, et al., "Improving Execution Concurrency of Large-scale Matrix Multiplication on Distributed Data-parallel Platforms,"
```
Good job with the references

coquelin77 replied:
thanks!

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -690,6 +690,74 @@ def __eq__(self, other):
 
+    def __matmul__(self, other):
+        """
+            Matrix multiplication of two DNDarrays
```
Identation is off

Markus-Goetz commented on heat/core/dndarray.py:
```diff
@@ -1113,6 +1181,11 @@ def __getitem__(self, key):
                     key[self.split] = key[self.split] - chunk_start
                     arr = self.__array[tuple(key)]
                     gout = list(arr.shape)
+                elif key[self.split] < 0 and self.gshape[self.split] + key[self.split] in range(chunk_start, chunk_end):
```
Is this a bug fix from another issue? If so, please reference it so that we can close it

coquelin77 replied:
fixes #293

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+    ----------
+    a : ht.DNDarray
+        2 dimensional: L x P
+
```
Extra line

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+
+    b : ht.DNDarray
+        2 dimensional: P x Q
+
```
Extra line

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+    out : ht.tensor
+        Optional
+        output tensor
+
```
Extra line

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+                  [11., 12., 13.],
+                  [12., 13., 14.]])
+    """
+    if a.gshape[-1] != b.gshape[-2]:
```
What about broadcasting, i.e. vector-matrix-mult?

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+        elif a.split == 1 and b.split == 0:
+            split_10_flag = True
+        else:
+            raise NotImplementedError("splits > 1 not implemented")
```
single quotation marks

Markus-Goetz commented on heat/core/linalg.py:
```diff
@@ -1,15 +1,459 @@
+        rem_map = factories.zeros((a.comm.size, 2, 2))
+        rem_map[a.comm.rank, 0, :] = (rem_a_out, rem_a)
+        rem_map[a.comm.rank, 1, :] = (rem_b, rem_b_out)
+        a.comm.Allreduce(MPI.IN_PLACE, rem_map, MPI.SUM)
```
Can we do an Asynchronous allreduce here, i.e. Iallreduce() and wait for the request later on? Same goes for l. 176 an so one. This way we could overlap communication and computation slightly.

Markus-Goetz replied:
The wait would happen on line 188?!

coquelin77 commented:
this also fixes #293

coquelin77 commented:
addressed #285 in most recent commit (54f9c46) @krajsek

coquelin77 commented:
> got some type error:
> 
> import heat as ht
> import numpy as np
> aa = ht.array(np.ones((4,1)), dtype=float)
> bb = ht.array(np.ones((1,4)), dtype=float)
> cc = aa @ bb
> 
> File "/home/heat/heat/core/dndarray.py", line 755, in **matmul**
> return linalg.matmul(self, other)
> File "/home/heat/heat/core/linalg.py", line 95, in matmul
> c._DNDarray__array[slice_0.start:slice_0.stop, :] += a._DNDarray__array @ b._DNDarray__array
> RuntimeError: output with type torch.FloatTensor doesn't match the desired type torch.DoubleTensor

is this the most recent version? line 95 is only an 'else' 
also i cannot reproduce this issue with the most recent commit

krajsek commented:
Sorry, please try this:

import heat as ht
import numpy as np
aa = ht.array(np.ones((4,1)))
bb = ht.array(np.ones((1,4)))
cc = aa @ bb

coquelin77 commented:
> Sorry, please try this:
> 
> import heat as ht
> import numpy as np
> aa = ht.array(np.ones((4,1)))
> bb = ht.array(np.ones((1,4)))
> cc = aa @ bb

this appears to be a bug in the conversion of the numpy tensor from a double to a float. when dtype.torch_type() is called it returns torch.float64. This conversion does not ever happen. This means that the dtype of the heat array is set to a float but the local tensors are still a DoubleTensor. I am opening up an issue now and will also attempt to fix it.

krajsek commented:
What is the general rule for the split dimension of the multiplied tensor?

coquelin77 commented:
> What is the general rule for the split dimension of the multiplied tensor?

The split dimension of the returned array is typically the split dimension of a. However, if a.split = None then the the c.split will be set as the split dimension of b. If both are None then c.split is also None. If a is a split vector then the returned vector will be of shape (1xQ) and will be split in the 1st dimension. If b is a vector and either a or b is split, then the returned vector will be of shape (Lx1) and will be split in the 0th dimension.

Markus-Goetz commented:
Bump, merge conflict with master
